File: src/com/mantono/ghapic/RequestConsumer.java
package com.mantono.ghapic;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Callable;

import javax.net.ssl.HttpsURLConnection;

public class RequestConsumer implements Callable<Response>
{
	private final Request request;
	private final String accessToken;

	public RequestConsumer(final String token, final Request request)
	{
		this.accessToken = token;
		this.request = request;
	}

	@Override
	public Response call() throws Exception
	{
		HttpsURLConnection connection = null;
		try
		{
			connection = (HttpsURLConnection) request.getUrl().openConnection();
			connection.setRequestMethod(request.getMethod().name());
			connection.setRequestProperty("Authorization", "token " + accessToken);
			connection.setRequestProperty("User-Agent", "ghapic");
			connection.setRequestProperty("Accept", "application/vnd.github.v3.text+json");
			connection.setDoOutput(true);
			
			connection.connect();
			
			final Map<String, List<String>> header = extractHeader(connection);
			final List<String> body = parseBody(connection);
			return new Response(header, body);
		}
		finally
		{
			if(connection != null)
			{
				connection.disconnect();
			}
		}
	}

	private List<String> parseBody(HttpsURLConnection connection)
	{
		
		try(InputStream is = connection.getInputStream();)
		{
			BufferedReader rd = new BufferedReader(new InputStreamReader(is));
			List<String> body = new ArrayList<String>(); 
			String line;
			while((line = rd.readLine()) != null)
				body.add(line);

			return body;
		}
		catch(IOException e)
		{
			e.printStackTrace();
		}
		return null;
	}

	private Map<String, List<String>> extractHeader(HttpsURLConnection connection)
	{
		return connection.getHeaderFields();
	}
}

File: src/main/kotlin/com/mantono/ghapic/RequestConsumer.kt
package com.mantono.ghapic

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import java.util.concurrent.Callable
import javax.net.ssl.HttpsURLConnection

const val VERSION: String = "0.1" // TODO read this from build.gradle defined version instead
private val mapper: ObjectMapper = ObjectMapper()

class RequestConsumer(val token: AccessToken, val request: Request): Callable<Response>
{
	override fun call(): Response
	{
		val connection: HttpsURLConnection =  request.resource.openConnection() as HttpsURLConnection
		return connection.run {
			requestMethod = request.method.name
			setRequestProperty("Authorization", "token " + token.token)
			setRequestProperty("User-Agent", "Ghapic/v$VERSION")
			setRequestProperty("Accept", "application/vnd.github.v3.text+json")
			doOutput = true

			connect()
			val body: JsonNode = mapper.readTree(inputStream)
			Response(headerFields, body)
		}
	}
}